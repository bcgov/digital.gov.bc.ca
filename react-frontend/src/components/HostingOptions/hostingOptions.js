import React from 'react';
import { Col, Row } from 'react-flexbox-grid';
import * as Scroll from 'react-scroll';

import DocumentTitle from 'react-document-title';
import InfoCard from '../Home/infocard';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faExternalLinkAlt } from '@fortawesome/free-solid-svg-icons';

import { PageContainer } from '../StyleComponents/pageContent';
import { HrefLink, HrefLinkScrollTo } from '../StyleComponents/htmlTags';
import { hostingOptionUrls } from '../../constants/urls';
import { Title, Heading, SubSubHeading } from '../StyleComponents/headings';

import HADSpdf from '../../documents/Hosting-and-Application-Development-Strategy.pdf';

function HostingOptions() {
  const ScrollElement = Scroll.Element;

  return (
    <DocumentTitle title="Hosting Options - Digital Government - Province of British Columbia">
      <div>
        <PageContainer>
          <Row middle="xs">
            <Col xs={12} lg={8}>
              <Title>Hosting Options</Title>
              <p>
                Before you create your awesome digital service, you should
                consider where it will live. The term “hosting” refers to the
                collection of infrastructure technologies that provide a
                foundation for the digital service.
              </p>
              <p>What is the solution you’re looking to deliver? </p>
              <ul>
                <li>
                  A modern web-based application? Use{' '}
                  <HrefLinkScrollTo
                    to="containers"
                    spy={true}
                    smooth={true}
                    offset={-180}
                  >
                    containers
                  </HrefLinkScrollTo>
                </li>
                <li>
                  Commercial off-the-shelf software? Use{' '}
                  <HrefLinkScrollTo
                    to="virtualServers"
                    spy={true}
                    smooth={true}
                    offset={-180}
                  >
                    virtual servers
                  </HrefLinkScrollTo>
                </li>
                <li>
                  Big data warehouse for hosting and working with massive
                  amounts of data? Use{' '}
                  <HrefLinkScrollTo
                    to="physicalServers"
                    spy={true}
                    smooth={true}
                    offset={-180}
                  >
                    physical servers
                  </HrefLinkScrollTo>
                </li>
              </ul>
              <p>
                The B.C. government has several hosting options, which are
                further described below. If you’re not sure, contact{' '}
                <HrefLink href="mailto:CITZAS@gov.bc.ca">
                  CITZAS@gov.bc.ca
                </HrefLink>{' '}
                to discuss the option that best fits your needs.
              </p>
            </Col>
            <Col xs={12} lg={4}>
              <InfoCard
                title={'IM/IT Standards'}
                description="A vision and approach for modernizing government’s efforts to provide excellent digital services by using best practices."
                linkText="Read the standards"
                linkPath={HADSpdf}
                height="300px"
              />
            </Col>
          </Row>
          <Row>
            <Col xs={12}>
              <ScrollElement name="containers" className="element" />
              <Heading>Containers </Heading>
              <SubSubHeading>Overview</SubSubHeading>
              <p>
                Containers provide a way for several digital services to share a
                single operating system. Digital teams can focus their efforts
                on the service they’re providing rather than installing or
                maintaining the infrastructure. Containers also support the idea
                of “infrastructure as code”, which means that a service can
                increase or decrease the resources it requires based on demand
                or even launch new instances of itself if something goes wrong.
              </p>
              <SubSubHeading>Pros </SubSubHeading>
              <ul>
                <li>Less maintenance</li>
                <li>Lower costs </li>
                <li>Most resilient of available hosting options </li>
                <li>
                  Promotes good practices such as automation and portability{' '}
                </li>
              </ul>
              <SubSubHeading>Cons </SubSubHeading>
              <ul>
                <li>
                  Steep learning curve, particularly among traditional teams{' '}
                </li>
                <li>
                  Requires increased diligence around how the service is
                  designed to take full advantage of built-in resilience of
                  containers{' '}
                </li>
                <li>
                  Limited to systems able to use the Linux operating system,
                  does not work well for Windows Server or large databases{' '}
                </li>
                <li>Emerging service with evolving support </li>
              </ul>

              <SubSubHeading>Recommendation</SubSubHeading>
              <p>
                Containers are the recommended hosting option for new digital
                services.{' '}
              </p>
              <SubSubHeading>How to onboard </SubSubHeading>
              <p>
                The container hosting options for the B.C. government are
                evolving and include public and private cloud platforms. Contact
                the{' '}
                <HrefLink href="mailto:pathfinder@gov.bc.ca">
                  platform services team
                </HrefLink>{' '}
                to discuss the best option for your service.
              </p>
              <ScrollElement name="virtualServers" className="element" />
              <Heading>Virtual servers </Heading>
              <SubSubHeading>Overview</SubSubHeading>
              <p>
                A virtual server provides a way for several digital services to
                share a single physical server. Virtual servers are an
                established hosting option for the B.C. government. They are
                more expensive than containers and require additional
                maintenance.
              </p>
              <SubSubHeading>Pros </SubSubHeading>
              <ul>
                <li>Supports Windows Server </li>
                <li>More resilient than physical servers </li>
                <li>Established service with well-documented support</li>
              </ul>
              <SubSubHeading>Cons </SubSubHeading>
              <ul>
                <li>
                  Requires ongoing management and upgrades to operating system
                </li>
              </ul>

              <SubSubHeading>Recommendation</SubSubHeading>
              <p>
                Virtual servers are recommended for off-the-shelf software and
                small/medium databases.{' '}
              </p>
              <SubSubHeading>How to onboard</SubSubHeading>
              <p>
                The B.C. government offers virtual servers in its data centers
                via the{' '}
                <HrefLink href={hostingOptionUrls.iStore}>
                  iStore order management tool
                  <FontAwesomeIcon
                    icon={faExternalLinkAlt}
                    style={{ paddingLeft: '5px' }}
                  />
                </HrefLink>
                .{' '}
              </p>
              <p>
                Virtual servers are also available to early adopters via the{' '}
                <HrefLink href="mailto:Cloud.Questions@gov.bc.ca">
                  Cloud Pathfinder team
                </HrefLink>
                .
              </p>
              <ScrollElement name="physicalServers" className="element" />
              <Heading>Physical servers</Heading>
              <SubSubHeading>Overview</SubSubHeading>
              <p>
                Physical servers represent the oldest hosting option currently
                available for onboarding for B.C. government teams. When using
                physical servers, teams need to be mindful that there is
                significant maintenance overhead.
              </p>
              <SubSubHeading>Pros </SubSubHeading>
              <ul>
                <li>Established service with well-documented support</li>
                <li>
                  Simplified licensing for certain types of legacy software
                </li>
              </ul>
              <SubSubHeading>Cons </SubSubHeading>
              <ul>
                <li>Most expensive hosting option</li>
                <li>Least resilient hosting option</li>
              </ul>
              <SubSubHeading>Recommendation</SubSubHeading>
              <p>Physical servers are not recommended.</p>
              <SubSubHeading>How to onboard</SubSubHeading>
              <p>
                The B.C. government offers physical servers in its data centers
                via the{' '}
                <HrefLink href={hostingOptionUrls.iStore}>
                  iStore order management tool
                  <FontAwesomeIcon
                    icon={faExternalLinkAlt}
                    style={{ paddingLeft: '5px' }}
                  />
                </HrefLink>
                .
              </p>
            </Col>
          </Row>
        </PageContainer>
      </div>
    </DocumentTitle>
  );
}

export default HostingOptions;
