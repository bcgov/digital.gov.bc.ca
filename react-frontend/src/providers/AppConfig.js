import React, { Component } from 'react';

export const AppConfigContext = React.createContext();

export class AppConfig extends Component {
  // TODO: It may be best to move the config set state into
  // constructor to make sure the config is assigned
  constructor() {
    super();
    this.state = { strapiApiUrl: null };
  }

  componentDidMount() {
    const xhr = new XMLHttpRequest();
    xhr.open('GET', '/config.json');

    xhr.send();

    xhr.onload = () => {
      if (xhr.status !== 200) {
        // analyze HTTP status of the xhrs response
        // do something
      } else {
        this.setState(JSON.parse(xhr.response));
      }
    };
  }

  render() {
    return (
      <AppConfigContext.Provider value={{ state: this.state }}>
        {this.props.children}
      </AppConfigContext.Provider>
    );
  }
}
