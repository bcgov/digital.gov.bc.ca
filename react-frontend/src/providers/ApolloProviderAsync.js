import React, { useContext, useEffect, useState } from 'react';
import { ApolloProvider } from 'react-apollo';
import ApolloClient from 'apollo-boost';

import { AppConfigContext } from './AppConfig';

// The apollo provider implements the strapi configuration
// by pulling the backend url from a config file. This allows
// the deployment config file instead of being baked into
// the build config.

function ApolloProviderAsync(props) {
  const config = useContext(AppConfigContext);
  const strapiURL = config['state']['strapiApiUrl'];

  // using client as state prevents it from being initialized needlessly
  const [client, setClient] = useState(
    new ApolloClient({
      uri: 'http://localhost:3000/graphql',
    })
  );

  // The useEffect is only called when config changes
  useEffect(() => {
    if (strapiURL) {
      setClient(new ApolloClient({ uri: strapiURL }));
    }
  }, [config, strapiURL]);

  return <ApolloProvider client={client}>{props.children}</ApolloProvider>;
}

export default ApolloProviderAsync;
