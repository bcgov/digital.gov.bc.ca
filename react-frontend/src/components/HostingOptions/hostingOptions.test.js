import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import { AppConfig } from '../../providers/AppConfig';
import { BrowserRouter as Router } from 'react-router-dom';
import HostingOptions from './hostingOptions';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <AppConfig>
        <MockedProvider>
          <Router>
            <HostingOptions />
          </Router>
        </MockedProvider>
      </AppConfig>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
