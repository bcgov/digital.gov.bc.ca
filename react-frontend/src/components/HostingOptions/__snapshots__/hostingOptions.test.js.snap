// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="sc-hmjpBu sc-dtMiey hwLXKX unuLi horizontalAlignment pageContainer container"
  >
    <div
      className="row middle-xs"
    >
      <div
        className="col-xs-12 col-lg-8"
      >
        <h1
          className="sc-GEcJY buFiZz"
        >
          Hosting Options
        </h1>
        <p>
          Before you create your awesome digital service, you should consider where it will live. The term “hosting” refers to the collection of infrastructure technologies that provide a foundation for the digital service.
        </p>
        <p>
          What is the solution you’re looking to deliver? 
        </p>
        <ul>
          <li>
            A modern web-based application? Use
             
            <a
              className="sc-ehCIER fbHnsV"
              onClick={[Function]}
            >
              containers
            </a>
          </li>
          <li>
            Commercial off-the-shelf software? Use
             
            <a
              className="sc-ehCIER fbHnsV"
              onClick={[Function]}
            >
              virtual servers
            </a>
          </li>
          <li>
            Big data warehouse for hosting and working with massive amounts of data? Use
             
            <a
              className="sc-ehCIER fbHnsV"
              onClick={[Function]}
            >
              physical servers
            </a>
          </li>
        </ul>
        <p>
          The B.C. government has several hosting options, which are further described below. If you’re not sure, contact
           
          <a
            className="sc-jeqYYF zjmip externalLink"
            href="mailto:CITZAS@gov.bc.ca"
          >
            CITZAS@gov.bc.ca
          </a>
           
          to discuss the option that best fits your needs.
        </p>
      </div>
      <div
        className="col-xs-12 col-lg-4"
      >
        <div
          className="ant-card ant-card-bordered sc-hKwCoD fApqXv cardRound"
          data-testid="styled-card"
        >
          <div
            className="ant-card-body"
            style={Object {}}
          >
            <h5
              className="sc-egiSv kfuvTV cardTitle"
            >
              Hosting and Application Development Strategy
            </h5>
            <p
              className="sc-jrQzUz kjzYZb cardDescription"
            >
              A vision and approach for modernizing government’s efforts to provide excellent digital services by using best practices.
            </p>
            <div
              className="sc-crHlIS guojQY cardLink"
            >
              <a
                className="sc-eJwXpk hGQaAH externalLink"
                href="Hosting-and-Application-Development-Strategy.pdf"
                rel="noopener noreferrer"
                target="_blank"
              >
                Read the strategy
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-external-link-alt fa-w-16 "
                  data-icon="external-link-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={
                    Object {
                      "paddingLeft": "5px",
                    }
                  }
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-xs-12"
      >
        <div
          className="element"
          name="containers"
        />
        <h2
          className="sc-gsNhcj elzHSl heading"
          style={
            Object {
              "paddingTop": "44px",
            }
          }
        >
          Containers 
        </h2>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Overview
        </h4>
        <p>
          Containers provide a way for several digital services to share a single operating system. Digital teams can focus their efforts on the service they’re providing rather than installing or maintaining the infrastructure. Containers also support the idea of “infrastructure as code”, which means that a service can increase or decrease the resources it requires based on demand or even launch new instances of itself if something goes wrong.
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Pros 
        </h4>
        <ul>
          <li>
            Less maintenance
          </li>
          <li>
            Lower costs 
          </li>
          <li>
            Most resilient of available hosting options 
          </li>
          <li>
            Promotes good practices such as automation and portability
             
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Cons 
        </h4>
        <ul>
          <li>
            Steep learning curve, particularly among traditional teams
             
          </li>
          <li>
            Requires increased diligence around how the service is designed to take full advantage of built-in resilience of containers
             
          </li>
          <li>
            Limited to systems able to use the Linux operating system, does not work well for Windows Server or large databases
             
          </li>
          <li>
            Emerging service with evolving support 
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Recommendation
        </h4>
        <p>
          Containers are the recommended hosting option for new digital services.
           
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          How to onboard 
        </h4>
        <p>
          The container hosting options for the B.C. government are evolving and include public and private cloud platforms. Contact the
           
          <a
            className="sc-jeqYYF zjmip externalLink"
            href="mailto:pathfinder@gov.bc.ca"
          >
            platform services team
          </a>
           
          to discuss the best option for your service.
        </p>
        <div
          className="element"
          name="virtualServers"
        />
        <h2
          className="sc-gsNhcj elzHSl heading"
          style={
            Object {
              "paddingTop": "44px",
            }
          }
        >
          Virtual servers 
        </h2>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Overview
        </h4>
        <p>
          A virtual server provides a way for several digital services to share a single physical server. Virtual servers are an established hosting option for the B.C. government. They are more expensive than containers and require additional maintenance.
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Pros 
        </h4>
        <ul>
          <li>
            Supports Windows Server 
          </li>
          <li>
            More resilient than physical servers 
          </li>
          <li>
            Established service with well-documented support
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Cons 
        </h4>
        <ul>
          <li>
            Requires ongoing management and upgrades to operating system
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Recommendation
        </h4>
        <p>
          Virtual servers are recommended for off-the-shelf software and small/medium databases.
           
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          How to onboard
        </h4>
        <p>
          The B.C. government offers virtual servers in its data centers via the
           
          <a
            className="sc-jeqYYF zjmip externalLink"
            href="https://ssbc-client.gov.bc.ca/order/istore.htm"
          >
            iStore order management tool
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-external-link-alt fa-w-16 "
              data-icon="external-link-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "paddingLeft": "5px",
                }
              }
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </a>
          .
           
        </p>
        <p>
          Virtual servers are also available to early adopters via the
           
          <a
            className="sc-jeqYYF zjmip externalLink"
            href="mailto:Cloud.Questions@gov.bc.ca"
          >
            Cloud Pathfinder team
          </a>
          .
        </p>
        <div
          className="element"
          name="physicalServers"
        />
        <h2
          className="sc-gsNhcj elzHSl heading"
          style={
            Object {
              "paddingTop": "44px",
            }
          }
        >
          Physical servers
        </h2>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Overview
        </h4>
        <p>
          Physical servers represent the oldest hosting option currently available for onboarding for B.C. government teams. When using physical servers, teams need to be mindful that there is significant maintenance overhead.
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Pros 
        </h4>
        <ul>
          <li>
            Established service with well-documented support
          </li>
          <li>
            Simplified licensing for certain types of legacy software
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Cons 
        </h4>
        <ul>
          <li>
            Most expensive hosting option
          </li>
          <li>
            Least resilient hosting option
          </li>
        </ul>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          Recommendation
        </h4>
        <p>
          Physical servers are not recommended.
        </p>
        <h4
          className="sc-ctqRej hqcCI subSubHeading"
        >
          How to onboard
        </h4>
        <p>
          The B.C. government offers physical servers in its data centers via the
           
          <a
            className="sc-jeqYYF zjmip externalLink"
            href="https://ssbc-client.gov.bc.ca/order/istore.htm"
          >
            iStore order management tool
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-external-link-alt fa-w-16 "
              data-icon="external-link-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "paddingLeft": "5px",
                }
              }
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;
